Show the mixer who's boss 😎

johnnovak opened the pull request:
# Description

This is about refactoring the mixer code _aggressively_ to bring it into the 21st century, kicking and screaming 🙀 The mixer was full of old-school stuff that probably made sense in the very early 2000s but not so much anymore (e.g., fixed-point math stuff).

We've been eyeing the mixer with kcgen for a while now, so it definitely had it coming 😎 I'm far from being done with it yet, but this is more than enough for a single mega PR 😅 

Main changes:

- Disable `-Wsign-conversion`, now with feeling 😎 🤘🏻 (https://github.com/dosbox-staging/dosbox-staging/pull/3650/commits/b28919685b6ff4674130235288885b1025813c21; my [first attempt ](https://github.com/dosbox-staging/dosbox-staging/pull/3631)was only partially successful)
- De-uint the mixer (use plain `int`s everywhere where it makes sense)
- De-uint the AdLib Gold, GUS, OPL, PC Speaker Impulse audio devices
- Replace fixed-point math in the mixer with proper float calculations
- Replace the ad-hoc ring buffer implementation in the mixer with a re-usable `RingBuffer` container that wraps `std::array` (see what I did there?! 😆)
- Use iterators in the mixer as much as possible, replacing for loops with manually maintained indices (improves readability and less error-prone)
- Extend `AudioFrame` with operators and use these in the mixer as much as possible (improves readability and less error-prone)
- Remove some time-stretching trickery from the mixer that attempts to handle buffer underruns, but in practice, makes zero difference (this simplifies the code somewhat)
- Remove some really weird "non-accurate rendering mode" from the mixer that was always activate _except_ when audio capturing was enabled; it attempted to minimise dropouts but I have no idea how... Basically, it did nothing useful.
- Remove lots of weird "magical" heuristic stuff that I just don't see how it can do anything of value. These are separate commits for easy bisecting, but it's good to get rid of cruft that doesn't seem to do anything useful...
- Fix OPL fade-out bug (https://github.com/dosbox-staging/dosbox-staging/pull/3650/commits/642d3870d586fc4d790f057402e9aa63aafb4598)
- Lots of shadowing warning fixes (less error-prone)
- And many small things and improvements—see the individual commits for details

Fixes https://github.com/dosbox-staging/dosbox-staging/issues/2704 and similar debug build only crashes. E.g., when you put your machine to sleep then wake it up, usually you'll get a crash with the debug builds, at least on macOS, but release builds are fine. Same deal when plugging in your headphones on macOS—yep, that caused the debug build to crash 😐 (I think some of the ridiculous fixed-point math stuff caused it, some counter got "too big" and tripped some asserts or something).

Anyway, all these debug-mode-special crashes are now gone.


## Related issues

- https://github.com/dosbox-staging/dosbox-staging/pull/3631
- https://github.com/dosbox-staging/dosbox-staging/issues/2704

# Manual testing

Lots of regression testing 😄 I'll test it more during the next few weeks until it gets reviewed (I'm assuming it won't go quickly). 

Testing tips:

- Test with chorus & reverb enabled.
- Test using the fast-forward hotkey to make sure we're still getting the sped-up tape "chipmunk" effect 😎 
- Test Direct DAC by configuring **Alone in the Dark** for the "Sound Blaster (no DMA)" option
- Test buffer underrun situations haven't regressed by setting `blocksize = 128` and `prebuffer = 5` or something low enough.
- Test the `reverse` mixer command still does its thing.
- Test setting the gain on different mixer channels and the master channel still works.
- Smoke test some AdLib, AdLib Gold (**Dune** is the only game to test), GUS, PS1 Audio, PC Speaker, Tandy & Game Blaster games—make sure there is sound and it's not distorted or anything.
- Test with `sbfilter = modern` (legacy linear interpolation) and `sbfilter = auto` and different SB models (Speex resampling and/or zero-order-hold upsampling, dependent on the SB model).
- Confirm audio capturing still works.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

weirddan455 reviewed (commented):
I've gone through to 6f0337ca9ca9c776648d12093922cbad49be6672 - "Remove is_mixer_irq_important() stuff "

Looks good so far.  Just found a typo.  I'll review more tomorrow.

johnnovak commented:
@weirddan455 Added some testing ideas, but I'll do more testing anyway in the coming days. I've already tested most of the basic stuff I've listed there.

weirddan455 commented on src/hardware/mixer.cpp:
```diff
@@ -1052,17 +1050,18 @@ void MixerChannel::Mix(const int frames_requested)
-		frames_remaining = (frames_remaining >> FreqShift) +
-		                   ((frames_remaining & FreqMask) != 0);
+		auto frames_remaining = iceil(
+		        static_cast<float>(frames_needed - frames_done) + freq_add);
```
Again, a little skeptical about this change.  It seems like frames remaining, frames needed, frames done, and frequency should all be integers.  I don't know what the heck the old code was doing with that bit shifting stuff though.  This whole thing just seems kind of weird to me.

johnnovak replied:
The old code _did not_ use integers; it used [fixed point arithmetics](https://en.wikipedia.org/wiki/Fixed_point_(mathematics)) to do the same calculations I'm doing now directly with floats.

>  It seems like frames remaining, frames needed, frames done, and frequency should all be integers. 

They are. But `freq_add` wasn't, that was a fixed-point decimal number, with a fractional part, potentially. Now that's a float.

weirddan455 reviewed (commented):
Alright I've gone through it all now.  Here's a few more comments.

johnnovak commented on src/hardware/mixer.cpp:
```diff
@@ -2449,12 +2447,15 @@ static void handle_mix_no_sound()
 {
 	ZoneScoped;
-	memset(stream, 0, static_cast<size_t>(len));
+	memset(stream, 0, static_cast<size_t>(bytes_requested));
```
Sure.

johnnovak commented on src/hardware/mixer.cpp:
```diff
@@ -2448,15 +2448,12 @@ static void SDLCALL mixer_callback([[maybe_unused]] void* userdata,
 
 	auto reduce_frames = 0;
 
-	// Local resampling counter to manipulate the data when sending it off
-	// to the callback
-	constexpr auto IndexShiftLocal = 14;
-
-	auto index_add = (1 << IndexShiftLocal);
-	auto index     = (index_add % frames_requested) ? frames_requested : 0;
+	// Used for time-stretching the audio in the buffer overrun scenario.
+	float index     = {};
+	float index_add = {};
```
Yeah, I'm flip-flopping between init to `0` or `{}`. Probably better to explicitly say `0`, so changing.

johnnovak commented on include/math_utils.h:
```diff
@@ -319,5 +319,11 @@ template float remap<float>(const float in_min, const float in_max,
                               const double v);
-							
+
+// From: http://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2
+constexpr bool is_power_of_two(int v)
+{
+	return v && ((v & (v - 1)) == 0);
+}
```
Yeah, I just wanted to have something that works at compile-time. I'm spoiled by Nim where you can use a significant part of the standard library at runtime, including all math functions, file I/O, etc.

Your solution is cleaner, so changing to that one.

johnnovak replied:
Okay, unfortunately you suggestion doesn't seem to work:

<img width="1064" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/90dadef4-d67d-4d01-935b-a0c205f28daa">

Using your first suggestion then 🤷🏻

weirddan455 replied:
Yeah, I just tried compiling it and also get a compile error.  What C++ version are we on?  It says the count method is consexpr since C++ 23.  I suppose we can't use this if we're on an older version.

https://en.cppreference.com/w/cpp/utility/bitset/count

johnnovak replied:
> What C++ version are we on?

On C++20, but not all C++20 features are supported in all the compilers that we use. So we use as much as we can.

We're not using C++23 features (and shouldn't as it's too early for that).

weirddan455 replied:
Actually, found a better solution.  This one compiles for me and is simpler (named after the x86 instruction and in standard since C++20)

```cpp
constexpr bool is_power_of_two(unsigned int v)
{
	return std::popcount(v) == 1;
}
```

weirddan455 replied:
Re-opening so you can see my above comment.  Up to you if you don't want to use it though.

johnnovak replied:
Very nice find @weirddan455, I like it! Simpler == better, so changing.

shermp replied:
Just noticed this, what about `std::has_single_bit()`?

johnnovak replied:
Cheers, I can try that quickly. Hopefully it will be accepted by all our compiler targets 🤞🏻

weirddan455 replied:
Welcome to C++ where the standard has 5 different ways to do the same thing :wink:

johnnovak replied:
@weirddan455 

"C++ is like an octopus made by nailing extra legs onto a dog."

"The 11 in C++11 refers to the number of legs that have now been nailed onto the dog whilst attempting to build a better octopus."

Still beats Cobol, PHP, and JavaScript... although that doesn't say much 😎

johnnovak commented on include/ring_buffer.h:
```diff
@@ -0,0 +1,163 @@
+class RingBuffer {
+
+#if POWER_OF_TWO_ARRAY_SIZE
+	static_assert(is_power_of_two(N), "RingBuffer size must be power of two");
```
Yeah, I'll make in unsigned int or something. I can be generally useful, not just for `size_t` if it's in math utils.

johnnovak commented on src/hardware/mixer.cpp:
```diff
@@ -153,8 +154,8 @@ struct ChorusSettings {
-	std::array<AudioFrame, MixerBufferByteSize> aux_reverb = {};
-	std::array<AudioFrame, MixerBufferByteSize> aux_chorus = {};
+	RingBuffer<AudioFrame, MixerBufferByteSize> aux_reverb = {};
+	RingBuffer<AudioFrame, MixerBufferByteSize> aux_chorus = {};
```
It's tested thoroughly every time to listen to some audio 😎 🤘🏻 

But yeah, I think I'll add some unit tests for extra confidence and to document the behaviour. There are many ways a ring buffer can behave, this one is a very simple one that will happily overwrite "itself"—but that's on purpose, it's for fast DSP stuff, so no checks.

johnnovak commented on src/hardware/mixer.cpp:
```diff
@@ -2270,12 +2263,12 @@ MixerChannel::~MixerChannel()
 }
 
-static constexpr int calc_tickadd(const int freq)
+static constexpr float calc_frames_per_tick(const int freq_hz)
```
You need fractional numbers if you're supporting _any_ sample rate. Sure, 44100 Hz or 48000 Hz is divisible by 1000, and you get an integer. But 22050 / 1000 = 22.05, and 11025 / 1000 = 11.25.

Currently you can also set any sample rate. `rate = 23456` will just work, try it. It you clamped the frames per tick to an integer here, that would cause issues.

Sure, using a fictional 23456 sample rate is pointless, but it works due to SDL's extra resampling step behind the scenes. Also, in my next PR I'm gonna relax the `rate` setting completely, so you can specify the 49716 OPL rate, or the 32000 MT-32 and SC-55 rate. Runnin the mixer at these weird rates are useful when you want to capture a bit-identical stream without any extra resampling for debugging and other such purposes when you need a WAV with the _exact_ generated sample stream. So I'm gonna allow it (with a warning in the logs).

All in all, it's much preferable to have a mixer that supports _any_ sample rate (within sane limits), and that's how it always has been.

Also, in general, it's best to use floats all the way throughout in DSP code, then convert to int where you really have to at the "boundaries". Internally, you often need to multiply numbers, etc., so it's rather pointless (and inefficient) to store something that can only be an integer as an `int` instead of just uniformly using `floats`. Converting between ints and floats has a performance impact too, and it's inconvenient to cast.

But, again, here we absolutely need fractional numbers.

johnnovak commented on src/dos/cdrom.h:
```diff
@@ -321,7 +321,7 @@ class CDROM_Interface_Image final : public CDROM_Interface
 		uint32_t                 startSector        = 0;
 		uint32_t                 totalRedbookFrames = 0;
-		int16_t buffer[MixerBufferByteSize * REDBOOK_CHANNELS] = {0};
+		int16_t buffer[MixerBufferByteSize]         = {};
```
Okay, I think you're right. I _assume_ the code was sane, that was my error 😓 

So this is how it looked like originally:

```cpp
int16_t buffer[MixerBufferLength * REDBOOK_CHANNELS] = {0}
```

I thought whoever wrote this assumed `MixerBufferLength` is in frames, not bytes, then multiplying by `REDBOOK_CHANNELS` (which equals to 2) would make completely sense.

But then, the code just passes raw pointers around to `buffer` like no big deal, and while there are some assert on the upper limit on the length of the data that can be written to `buffer` in the `decode` method, that limit could vastly exceed the actual size of the array.

For the record `MixerBufferByteSize` is 16384 (16k). So I'll just change this to `MixerBufferByteSize * 2` and put a comment there that the CD-ROM audio code is unsafe, might only not blow up by fluke, and we should really change this to a self-sizing `std::vector` sometime.

Good pickup!

johnnovak commented:
Thanks for the thorough review @weirddan455, much appreciated 🎉 I've addressed all your comments except for adding unit tests for the RingBuffer. I'll do that in the coming days plus I'll retest all audio devices one more time, then it's merge time!

weirddan455 commented:
I was more thinking of manual testing (which it'll get plenty of naturally by the time of next release).  But sure if you want to write some unit tests, that's up to you.

johnnovak commented:
Okay @weirddan455, I've added RingBuffer unit tests which has caught some bugs around the `+=` && `-=` operators! Crikey!

https://github.com/dosbox-staging/dosbox-staging/pull/3650/commits/9e930185b3b955f41d36356fb98435c93fae9a5d

Other than that, I'm done. I've also done yet another round of testing, this is as good as it gets from my POV.

johnnovak commented:
Thanks for the review @weirddan455
